universal-application-tool-0.0.1/app/views/applicant/ApplicantProgramInfoView.java

```diff
@@ -0,0 +1,153 @@
+    String programInfo = program.localizedDescription().getOrDefault(preferredLocale);
+
+    HtmlBundle bundle =
+        layout
+            .getBundle()
+            .addMainStyles(Styles.MX_12, Styles.MY_8)
+            .addMainContent(topContent(programTitle, programInfo, messages))
+            .addMainContent(createButtons(applicantId, program.id(), messages));
+
+    return layout.renderWithNav(request, userName, messages, bundle);
+  }
+
+  private ContainerTag topContent(String programTitle, String programInfo, Messages messages) {
+    String programsLinkText = messages.at(MessageKey.TITLE_PROGRAMS.getKeyName());
+    String homeLink = routes.HomeController.index().url();
+    ContainerTag allProgramsDiv =
+        a("<")
+            .withHref(homeLink)
+            .withClasses(Styles.TEXT_GRAY_500, Styles.TEXT_LEFT)
+            .with(span().withText(programsLinkText).withClasses(Styles.PX_4));
+
+    ContainerTag titleDiv =
+        h2().withText(programTitle)
+            .withClasses(
+                BaseStyles.TEXT_SEATTLE_BLUE,
+                Styles.TEXT_2XL,
+                Styles.FONT_SEMIBOLD,
+                Styles.TEXT_GRAY_700,
+                Styles.MT_4);
+
+    // "Markdown" the program description.
+    String[] lines = Iterables.toArray(Splitter.on("\n").split(programInfo), String.class);
+    ImmutableList<DomContent> items = formatText(lines, false);
+    ContainerTag descriptionDiv = div().withClasses(Styles.PY_2).with(items);
+
+    return div(allProgramsDiv, titleDiv, descriptionDiv);
+  }
+
+  /** Adds the ability to create accordions and lists from data in text fields. */
+  private ImmutableList<DomContent> formatText(String[] lines, boolean preserveEmptyLines) {
+    ImmutableList.Builder<DomContent> builder = new ImmutableList.Builder<DomContent>();
+    for (int i = 0; i < lines.length; i++) {
+      String line = lines[i].trim();
+      if (line.startsWith("###")) { // We're calling this an accordion.
+        String accordionHeader = line.substring(3);
+        int next = i + 1;
+        ArrayList<String> content = new ArrayList<>();
+        while (next < lines.length && lines[next].startsWith(">")) {
+          content.add(lines[next].substring(1));
+          next++;
+        }
+        i = next - 1;
+        builder.add(buildAccordion(accordionHeader, content));
+      } else if (line.startsWith("*")) { // unordered list item.
+        ArrayList<String> items = new ArrayList<>();
+        items.add(line.substring(1));
+        int next = i + 1;
+        while (next < lines.length && lines[next].startsWith("*")) {
+          items.add(lines[next].substring(1));
+          next++;
+        }
+        i = next - 1;
+        builder.add(buildList(items));
+      } else if (line.length() > 0) {
+        ImmutableList<DomContent> lineContent = createLinksAndEscapeText(line);
+        builder.add(div().with(lineContent));
+      } else if (preserveEmptyLines) {
```

Comment:
(blocking) this should have an `else`.

Reply:
The `else` here is implicitly "if we're not preserving empty lines, throw this away.'

Reply:
It might be more readable if there was a method like `lineToElementEnum(line)` that returned a typed enum, and this used a switch statement with an empty `default`.

Reply:
IMO this should be a non-blocking suggestion. We're pretty short on time these days.

Reply:
Yeah, I like the idea of making an enum and turning the long `if/else if * 4` block into a switch statement, but Bion's right... that's quite a bit of extra effort for a feature that I feel like I'm sneaking in at the 11th (maybe only 10th) hour.